1-Draft/RFC00XX-Update-Error-View.md

```diff
@@ -0,0 +1,181 @@
+---
+RFC: Update $ErrorView with Single String Error Messages
+Author: Jason Helmick
+Status: Draft
+Version: 0.0.0
+Area: PowerShell
+Comments Due: 
+---
+
+# Update $ErrorView with Single String Error Messages
+
+When an error occurs in PowerShell, the customers on-screen error message experience currently
+provides a level of detail that obscures the inner exception message from being read by
+new or occasional PowerShell users.
+To improve both comprehension and the troubleshooting experience, a single string of text
+containing the error will be displayed.  A cmdlet will provide the full error details when desired.
+
+## Motivation
+
+The on-screen experience, when receiving an error message,
+is controlled through the views NormalView (the default) and CategoryView. These are user selectable though the preference variable $ErrorView.
+The addition of a “ConciseView” as default that contains only specific error
+information producing a single line on screen and in logs will improve customer success.
+
+Comprehensive error information is available to the customer through
+accessing the error object. For new and occasional customers,
+a cmdlet “Resolve-ErrorRecord” will simplify the process of getting detailed error information, improving customers success.
+
+## Specification
+
+The proposal is to add a new single-line default error message view called ConciseView
+and add a new cmdlet (Resolve-ErrorRecord) to provide detailed error information.
+
+__Key Design Considerations__
+
+1. To reduce confusion and improve debugging success,
+error messages by default should produce a single line of text, including the word “ERROR:” to make consistent with Verbose and Warning messages. This is added to $errorView as view named "ConciseView".
+
+- $ErrorView should contain these views
+
+    + NormalView
+    + CategoryView
+    + ConciseView
+
+- Error Message syntax for ConciseView
+
+```powershell
+ERROR: <Exception Message>: <Positional Info>
+ERROR: Cannot find path ‘C:\blah’ because it does not exist: At line:1 char:1
+```
+
+2. A new cmdlet “Resolve-ErrorRecord” will produce detailed error information similar to Format-List * -Force.
+
```

Comment:
Why not just do this instead:

1. Change PowerShell so that errors _and_ exceptions can be piped to `Format-List` (or other `Format-*` cmdlets) _without_ needing `-Force`.
1. If you really feel $error is not easy enough to use, provide a `Get-Error` cmdlet that allows you to specify `-First`, `-Last`, etc. parameters as well as a `-View` parameter to let you see the view you want without changing your default view, so instead of `Get-LastError`, PowerShell could offer `Get-Error -Last` (plus the other options available via the other parameters, of course).

Reply:
I would like to understand the historical reason to require `-Force`, I find it quite annoying myself.  cc @JamesWTruher @BrucePay

Reply:
FWIW, in my FormatPx module I made this change several years ago so that errors and exceptions can pipe to `Format-*` cmdlets without `-Force` and give you the format you want, and haven't heard a complaint or run into any problems with that approach since I first implemented it.

Reply:
One more thing on the topic of `Format-*` cmdlets. We're talking about views for objects in PowerShell here. These views are presented by the F&O layer of PowerShell. Yet I can't just call up the view on demand, even if I `-Force` it, and instead have to specifically muck around with the `$ErrorView` variable. This feels sloppy and inconsistent with how formatting is supposed to work in PowerShell.

Here's what I'm talking about, if that didn't make any sense to you:

![image](https://user-images.githubusercontent.com/3515485/62319549-975a6300-b474-11e9-8238-06997b7768e4.png)

Reply:
As script writer and developer I'd want to have full error view (including InnerException) by default. Can we have a feature to switch default format view?

Reply:
Exposing the error views as named "Custom" formats would make a lot more sense than what we have now...

Reply:
> As script writer and developer I'd want to have full error view (including InnerException) by default. Can we have a feature to switch default format view?

@iSazonov: The `$ErrorView` variable provides this functionality.

In fact, that is the only benefit to having an `$ErrorView` variable at all, as far as I can tell: to allow users to change the default view, which is something that you can't normally do in PowerShell's F&O layer. Unfortunately, that's specific to just errors (i.e. you can't change default views for other object types without defining new formats and pre-pending them). Nor can you temporarily change how F&O works for errors as they occur (i.e. this still shows the error in normal view: `& {try {$global:ErrorView = 'CategoryView'; 1/0} finally {$global:ErrorView = 'NormalView'}}`).

For now leaving `$ErrorView` as a way to change the default error view feels like the right thing to do, with a separate RFC to allow for setting the default view for any object, if there's sufficient need for that in types other than `ErrorRecord` (I don't know that there is).

Aside: I just discovered that `Get-FormatData` shows format data for all objects _except_ those that use a custom control for formatting. Intentional feature, or bug?

Reply:
I'm leaning towards your point that it feels like the right thing to do now, and if someone really wants the old error messages by default, its fairly easy and plenty of reference documentation for $errorview.

Reply:
@KirkMunro Thanks for clarify! 
> Aside: I just discovered that Get-FormatData shows format data for all objects except those that use a custom control for formatting. Intentional feature, or bug?

It is not "aside". It is key question because
> you can't change default views for other object types without defining new formats and pre-pending them

 \- I believe this is more right direction than partial fix for error formatting.

Reply:
> I would like to understand the historical reason to require `-Force`, I find it quite annoying myself. cc @JamesWTruher @BrucePay

@SteveL-MSFT Did you ever hear from Jim Truher or Bruce Payette on the historical reasoning behind this?

Reply:
No, let me bring it up at next Committee meeting.

Reply:
> Why not just do this instead:
> 2. If you really feel $error is not easy enough to use, provide a `Get-Error` cmdlet 

Because powershell will treat error as an implied alias for get-error you can just type "error" for the last error details.  or your can do $error[2] | format-* for something specific.  I guess it would have an alias out of the box, but why type resolve-error instead of $error[0]

Reply:
RE @KirkMunro if we have cmdlets that will display additional error info, please, let's not display that additional info with the error fg color.  Using the error color is useful for drawing users' attention to errors in the console or a log file.  But once I'm (or a script) is asking for more error information, I don't believe it needs to be displayed in red. And I suspect that reading lots of "red on black" text is harder versus using the standard fg/bg colors.

Reply:
I agree with that @rkeithhill, and I think that's the way it works already in the PS 7 preview, with errors showing with color as they occur, but captured errors (`-ErrorVariable) or historical errors in `$error` not showing colors. Still, its worth calling out so that we don't put the colors back in.

The one place where I still see colored errors that puzzles me a little is redirected errors. For example, if I invoke `$r = gps -id 12345678 *>&1; $r`, I see colored errors. I tend to think "error" text (which sometimes is simply text sent to stderr in certain command line tools) that is redirected to standard out should just render like any other standard output. That's not specific to this RFC, but I thought I'd share the thought while we were talking about error colors.

Reply:
I support the idea of reducing the amount of "error color" text on the console to a minimum. Today when something goes wrong there is just too much text shown in the error color, with only a small part that conveys useful information to interpret the error.

I also support outputting the stderr from external commands in the default, not error color.

Reply:
Why is this marked as resolved without changes or comments responding to the [original post](https://github.com/PowerShell/PowerShell-RFC/pull/215#discussion_r309509455)?